src/restruct_poc/constraints/calculate_constraint_residual_gradient.hpp

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <Kokkos_Core.hpp>
+
+#include "calculate_cylindrical_constraint.hpp"
+#include "calculate_fixed_bc_constraint.hpp"
+#include "calculate_prescribed_bc_constraint.hpp"
+#include "calculate_rigid_constraint.hpp"
+#include "calculate_rotation_control_constraint.hpp"
+#include "constraints.hpp"
+
+#include "src/restruct_poc/math/quaternion_operations.hpp"
+#include "src/restruct_poc/math/vector_operations.hpp"
+#include "src/restruct_poc/types.hpp"
+
+namespace openturbine {
+
+struct CalculateConstraintResidualGradient {
+    Kokkos::View<ConstraintType*>::const_type type;
+    Kokkos::View<size_t* [2]>::const_type node_index;
+    Kokkos::View<size_t* [2]>::const_type row_range;
+    Kokkos::View<size_t* [2][2]>::const_type node_col_range;
+    Kokkos::View<double* [3]>::const_type X0_;
+    Kokkos::View<double* [3][3]>::const_type axis;
```

Comment:
@ddement changing `x_axis`, `y_axis`, and `z_axis` to a 3x3 matrix without comments makes it very hard to know what data is in there. The name `axis` implies a 3 element vector. This change significantly reduces the readability of this code and makes it harder to understand what is happening. And yes, I see that the matrix is unpacked into vectors before use, but it's just another layer of obscurity when somebody is passing a view called `axis` into `CalculateConstraintResidualGradient`. At the very least this should be renamed to `axes` since it is a view of multiple axes.

Reply:
I agree renaming it will make things clearer, I'll do that.  

I packed them into a single 3x3 entry because we typically want all three of them when we want one of them.  This not only simplifies calling kernels that use them, but it also improves caching/coalescing behavior.